saleor/order/tasks.py

```diff
@@ -117,3 +123,30 @@ def expire_orders_task():
     now = timezone.now()
     manager = get_plugins_manager()
     _expire_orders(manager, now)
+
+
+@app.task
+def delete_expired_orders_task():
+    now = timezone.now()
+
+    channel_qs = Channel.objects.filter(
+        delete_expired_orders_after__gt=timedelta(),
+        id=OuterRef("channel"),
+    )
+
+    qs = Order.objects.annotate(
+        delete_expired_orders_after=Subquery(
+            channel_qs.values("delete_expired_orders_after")[:1]
+        )
+    ).filter(
+        ~Exists(TransactionItem.objects.filter(order=OuterRef("pk"))),
+        ~Exists(Payment.objects.filter(order=OuterRef("pk"))),
+        expired_at__isnull=False,
+        status=OrderStatus.EXPIRED,
+        expired_at__lte=now - F("delete_expired_orders_after"),  # type:ignore
+    )
+    ids_batch = qs.values_list("pk", flat=True)[:DELETE_EXPIRED_ORDER_BATCH_SIZE]
+
+    Order.objects.filter(id__in=ids_batch).delete()
+    if qs.exists():
```

Comment:
I wonder if that's necessary? It performs an extra query just to run it again in next recursion, maybe that can be
```
if not ids_batch:
    return
Order.objects.filter(id__in=ids_batch).delete()
delete_expired_orders_task.delay()
```
to avoid calling `.exists()` and evaluating query again?

Reply:
Good catch! thanks